pkg/compression/compression.go

```diff
@@ -35,32 +36,85 @@ func XzDecompressor(r io.Reader) (io.ReadCloser, error) {
 	return ioutil.NopCloser(r), nil
 }
 
+// compressorFunc returns the compressed stream, given a compressed stream.
+// The caller must call Close() on the compressed stream (even if the compressed input stream does not need closing!).
+type compressorFunc func(io.Writer, int) (io.WriteCloser, error)
+
+// GzipCompressor is a CompressorFunc for the gzip compression algorithm.
+func GzipCompressor(r io.Writer, level int) (io.WriteCloser, error) {
+	if level > 0 {
+		return pgzip.NewWriterLevel(r, level)
+	}
+	return pgzip.NewWriter(r), nil
+}
+
+// Bzip2Compressor is a CompressorFunc for the bzip2 compression algorithm.
+func Bzip2Compressor(r io.Writer, level int) (io.WriteCloser, error) {
+	return nil, fmt.Errorf("bzip2 compression not supported")
+}
+
+// XzCompressor is a CompressorFunc for the xz compression algorithm.
+func XzCompressor(r io.Writer, level int) (io.WriteCloser, error) {
+	return xz.NewWriter(r)
+}
+
+const (
+	// Gzip is the gzip compression
+	Gzip = "gzip"
+	// Bzip2 is the bzip2 compression
+	Bzip2 = "bzip2"
+	// Xz is the xz compression
+	Xz = "xz"
+	// Zstd is the zstd compression
+	Zstd = "zstd"
+)
+
 // compressionAlgos is an internal implementation detail of DetectCompression
 var compressionAlgos = map[string]struct {
 	prefix       []byte
 	decompressor DecompressorFunc
 }{
-	"gzip":  {[]byte{0x1F, 0x8B, 0x08}, GzipDecompressor},                 // gzip (RFC 1952)
-	"bzip2": {[]byte{0x42, 0x5A, 0x68}, Bzip2Decompressor},                // bzip2 (decompress.c:BZ2_decompress)
-	"xz":    {[]byte{0xFD, 0x37, 0x7A, 0x58, 0x5A, 0x00}, XzDecompressor}, // xz (/usr/share/doc/xz/xz-file-format.txt)
+	Gzip:  {[]byte{0x1F, 0x8B, 0x08}, GzipDecompressor},                 // gzip (RFC 1952)
+	Bzip2: {[]byte{0x42, 0x5A, 0x68}, Bzip2Decompressor},                // bzip2 (decompress.c:BZ2_decompress)
+	Xz:    {[]byte{0xFD, 0x37, 0x7A, 0x58, 0x5A, 0x00}, XzDecompressor}, // xz (/usr/share/doc/xz/xz-file-format.txt)
+	Zstd:  {[]byte{0x28, 0xb5, 0x2f, 0xfd}, ZstdDecompressor},           // zstd (http://www.zstd.net)
 }
 
-// DetectCompression returns a DecompressorFunc if the input is recognized as a compressed format, nil otherwise.
+// compressors maps an algorithm to its compression function
+var compressors = map[string]compressorFunc{
```

Comment:
This could just as well be merged with `compressionAlgos` — but see the other comments about turning this into methods.